<a href="https://colab.research.google.com/github/ruderikissa/BEDU/blob/main/Final_proyect_Group6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<font size=12 color='dodgerblue'><center>Proyecto Integral Grupo 6</center></font>

---

##<font color='dodgerblue'>Integrantes:</font>

*   Alfredo Cuevas
*   Jorge Arsita
*   José María Montes Montiel
*   Rodrigo Garmendia

##<font color='dodgerblue'>Tema:</font><font color='crimson'> `Volcanes en el Mundo Tipo y Riesgo para la población`.</font>

![popocatepetl](https://www.montero.org.mx/assets/images/mountains/popocatepetl-centro-montanas.jpg)

###<font color='dodgerblue'>Fases:</font>
1. Identificación del tema a resolver o explorar.
2. Planteamiento de preguntas
3. Obtención de colección de datos
4. Análisis preliminar
5. Limpieza de datos
6. Transformación, filtrado y ordenamiento de datos
7. Conclusiones.

#<font size=6 color='dodgerblue'>1. Identificación del tema a resolver o explorar.</font>


---


En este caso se planteó el interés del equipo por **explorar** qué tipo de volcanes hay en el planeta y el riesgo que representan para la población en las diferentes zonas poblacionales.

Puede resultar para algunos grupos de la población mundial que la actividad volcánica solo sea parte de una noticia en medios de comunicaciones y redes sociales, pero para otra parte de la población es una situación de alto riesgo.

La intención de este proyecto es tener identificado la relación de volcanes clasificada por su riesgo de erupción, tipo de volcán, ubicación (región, subregión, país)

En particular el tema resulta de interés para el equipo por la intrincada situación geográfica y geológica de México, especialmente en el centro del país, donde el Eje Neovolcánico atraviesa el pais de costa a costa y coincide con la zona más densamente poblada de México, donde se encuentran grandes metrópolis como Guadalajara, Colima, Toluca, México y Puebla, entre otras.

Si bien no se tiene registrada alguna erupción catastrófica de estos volcanes en épocas recientes, no se puede ignorar el hecho de su actividad constante, en particular del Popocatépetl (Xihutépetl, Xalliquehuac, o Don Goyo) y del Volcán de Colima, con fumarolas (algunas muy vistosas y de gran tamaño), expulsión de material volcánico, explosiones y ruidos en el domo del volcán. Todo esto, resalta el hecho de que la Ciudad de México, Tlaxcala y Puebla podrían estar expuestas a un riesgo de gran magnitud si consideramos la cantidad de habitantes que tiene la Ciudad y las zonas cercanas al volcán en los estados circundantes.

Estas son algunas de las razones por las que se considera importante tener información relacionada con los volcanes, sus características y actividad.


#<font size=6 color='dodgerblue'>2. Planteamiento de preguntas</font>


---


Algunos de los cuestionamientos que se desean explorar son:
1.   ¿Cuáles volcanes representan mayor riesgo?
2.   ¿Qué país tiene el mayor riesgo potencial a una catástrofe volcánica?
3.   ¿De cuáles se tiene registro de -al menos una- erupción?
4.   ¿Qué tipos de volcanes existen?
5.   ¿Dónde se ubican? (uso de mapas interactivos)
6.   ¿Qué país/continente/región tiene la mayor cantidad de volcanes?
7.   Altitud media
8.   Relaciones tectónica de placas - tipo de volcán.




#<font size=6 color='dodgerblue'>3. Obtención de colección de datos</font>

---

Se obtuvo una colección de datos del sitio [volcano](https://volcano.si.edu/), el cual cuenta con información detallada de varios aspectos relacionados con los volcanes (boletines, reportes, bases de datos, galerías, etc.) que permiten obtener un compendio completo para explorar y responder los cuestionamientos planteados por el equipo.

Dentro del material encontrado se cuenta con dos bases de datos que corresponden a los periodos del pleistoceno (1.8 millones de años a 12,000 a.n.e. aproximadamente) y el holoceno (12,000 a.n.e.  a la actualidad)

#<font size=6 color='dodgerblue'>4. Análisis Exploratorio de Datos</font>

---

Se trabajó inicialmente con el [archivo holoceno original](https://raw.githubusercontent.com/acuevass/Proyecto_ProcesamientoDatosPython/main/Volcanes_holoceno_original.csv), dado que nuestro interés está en la actividad volcánica reciente y el archivo del pleistoceno recopila información de volcanes que llevan extintos más de 12,000 años. 

In [ ]:
# Se realizan las importaciones necesarias para iniciar la carga del dataset.
import pandas as pd
import numpy as np

Carga inicial del dataset Volcanes para obtener el DataFrame con el que trabajaremos en el proyecto.

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/acuevass/Proyecto_ProcesamientoDatosPython/main/Volcanes_holoceno.csv')

Se valida que se haya generado adecuadamente el DataFrame.

In [ ]:
df.head()

Volcano Number               Volcano Name  Country Primary Volcano Type  \
0          210010  West Eifel Volcanic Field  Germany              Maar(s)   
1          210020            Chaine des Puys   France         Lava dome(s)   
2          210030        Olot Volcanic Field    Spain  Pyroclastic cone(s)   
3          210040   Calatrava Volcanic Field    Spain  Pyroclastic cone(s)   
4          211004               Colli Albani    Italy              Caldera   

    Activity Evidence Last Known Eruption                          Region  \
0      Eruption Dated            8300 BCE  Mediterranean and Western Asia   
1      Eruption Dated            4040 BCE  Mediterranean and Western Asia   
2   Evidence Credible             Unknown  Mediterranean and Western Asia   
3      Eruption Dated            3600 BCE  Mediterranean and Western Asia   
4  Evidence Uncertain             Unknown  Mediterranean and Western Asia   

        Subregion  Latitude  Longitude  Elevation (m)  \
0  Western Europe    50.170       6.85            600   
1  Western Europe    45.775       2.97           1464   
2  Western Europe    42.170       2.53            893   
3  Western Europe    38.870      -4.02           1117   
4           Italy    41.730      12.70            949   

                 Dominant Rock Type  \
0                           Foidite   
1             Basalt / Picro-Basalt   
2  Trachybasalt / Tephrite Basanite   
3             Basalt / Picro-Basalt   
4                           Foidite   

                               Tectonic Setting  
0        Rift zone / Continental crust (>25 km)  
1        Rift zone / Continental crust (>25 km)  
2       Intraplate / Continental crust (>25 km)  
3       Intraplate / Continental crust (>25 km)  
4  Subduction zone / Continental crust (>25 km)

La forma de nuestro dataset es de 1,332 registros por 13 columnas:

In [ ]:
print('El tamaño de la base de datos es: ', df.shape)

El tamaño de la base de datos es:  (1332, 13)


Las columnas detectadas:

In [ ]:
print([col for col in df.columns])

['Volcano Number', 'Volcano Name', 'Country', 'Primary Volcano Type', 'Activity Evidence', 'Last Known Eruption', 'Region', 'Subregion', 'Latitude', 'Longitude', 'Elevation (m)', 'Dominant Rock Type', 'Tectonic Setting']


Detalle de las columnas:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1332 entries, 0 to 1331
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Volcano Number        1332 non-null   int64  
 1   Volcano Name          1332 non-null   object 
 2   Country               1332 non-null   object 
 3   Primary Volcano Type  1332 non-null   object 
 4   Activity Evidence     1331 non-null   object 
 5   Last Known Eruption   1332 non-null   object 
 6   Region                1332 non-null   object 
 7   Subregion             1332 non-null   object 
 8   Latitude              1332 non-null   float64
 9   Longitude             1332 non-null   float64
 10  Elevation (m)         1332 non-null   int64  
 11  Dominant Rock Type    1307 non-null   object 
 12  Tectonic Setting      1327 non-null   object 
dtypes: float64(2), int64(2), object(9)
memory usage: 135.4+ KB


Revisamos elementos nulos por columna.

In [ ]:
df.isna().sum()

Volcano Number           0
Volcano Name             0
Country                  0
Primary Volcano Type     0
Activity Evidence        1
Last Known Eruption      0
Region                   0
Subregion                0
Latitude                 0
Longitude                0
Elevation (m)            0
Dominant Rock Type      25
Tectonic Setting         5
dtype: int64

Y también checamos el tipo de dato por columna.

Vemos que la mayoría son del tipo object, dos del tipo float64 y uno del tipo int64.

In [ ]:
df.dtypes

Volcano Number            int64
Volcano Name             object
Country                  object
Primary Volcano Type     object
Activity Evidence        object
Last Known Eruption      object
Region                   object
Subregion                object
Latitude                float64
Longitude               float64
Elevation (m)             int64
Dominant Rock Type       object
Tectonic Setting         object
dtype: object

En un primer escrutinio se detectaron los siguientes incidentes:
- Los nombres de las columnas no están en formato **snake_case** de acuerdo a la convención de nombramiento en Python.
- La columna `Volcano Number` es un identificador y no aporta información relevante.
- La fecha de la última erupción `Last Know Eruption` está en tipo 'object'.
- Se encontraron algunos registros NaNs en los campos `Activity Evidence`, `Dominant Rock Type` y `Tectonic Setting`.
> Aunque no hay renglones completos de NaNs, sí se observó que en la columna `Last Known Eruption` existen valores de fecha **Unknow**. Esto es importante ya que parte de la propuesta de análisis tiene considera este dato.


Se observa que se cuenta con la información requerida para nuestro análisis y que debe ser tratada adecuadamente para la obtención de conclusiones.

El dataset contiene información relacionada a erupciones, características de los volcanes y su ubicación.

#<font size=6 color='dodgerblue'>5.Limpieza de datos</font>


---
La Limpieza de Datos nos permitió preparar adecuadamente el DataFrame para su análisis acorde al planteamiento de la exploración propuesta.
Permitiendo una obtención de información relevante sobre los datos.

##Normalización de nombres de columnas

Como se resaltó anteriormente los nombres de las columnas no siguen la convención adecuada para el ambiente de desarrollo con Python, por lo que es propio normalizarlos. Para ello nos apoyaremos de la biblioteca 're'.



In [ ]:
# Creamos un diccionario con llave=antiguo-nombre y valor=nuevo-nombre
import re
diccionario_convencion = {}
for col in [elem for elem in df.columns]:
  elem = re.sub(" ","_",col)
  elem = re.sub('\W','',elem)
  elem = elem.lower()
  diccionario_convencion[col] = elem

In [ ]:
# Validamos el diccionario creado.
diccionario_convencion

{'Volcano Number': 'volcano_number',
 'Volcano Name': 'volcano_name',
 'Country': 'country',
 'Primary Volcano Type': 'primary_volcano_type',
 'Activity Evidence': 'activity_evidence',
 'Last Known Eruption': 'last_known_eruption',
 'Region': 'region',
 'Subregion': 'subregion',
 'Latitude': 'latitude',
 'Longitude': 'longitude',
 'Elevation (m)': 'elevation_m',
 'Dominant Rock Type': 'dominant_rock_type',
 'Tectonic Setting': 'tectonic_setting'}

Efectuamos el renombramiento.

In [ ]:
df = df.rename(columns = diccionario_convencion)

Y validamos.

In [ ]:
df.columns

Index(['volcano_number', 'volcano_name', 'country', 'primary_volcano_type',
       'activity_evidence', 'last_known_eruption', 'region', 'subregion',
       'latitude', 'longitude', 'elevation_m', 'dominant_rock_type',
       'tectonic_setting'],
      dtype='object')

Procedemos a eliminar la primera columna para liberar memoria y por limpieza visual.

In [ ]:
df = df.drop('volcano_number', axis=1)

Se revisan nuevamente los cambios efectuados.

In [ ]:
df.head()

volcano_name  country primary_volcano_type  \
0  West Eifel Volcanic Field  Germany              Maar(s)   
1            Chaine des Puys   France         Lava dome(s)   
2        Olot Volcanic Field    Spain  Pyroclastic cone(s)   
3   Calatrava Volcanic Field    Spain  Pyroclastic cone(s)   
4               Colli Albani    Italy              Caldera   

    activity_evidence last_known_eruption                          region  \
0      Eruption Dated            8300 BCE  Mediterranean and Western Asia   
1      Eruption Dated            4040 BCE  Mediterranean and Western Asia   
2   Evidence Credible             Unknown  Mediterranean and Western Asia   
3      Eruption Dated            3600 BCE  Mediterranean and Western Asia   
4  Evidence Uncertain             Unknown  Mediterranean and Western Asia   

        subregion  latitude  longitude  elevation_m  \
0  Western Europe    50.170       6.85          600   
1  Western Europe    45.775       2.97         1464   
2  Western Europe    42.170       2.53          893   
3  Western Europe    38.870      -4.02         1117   
4           Italy    41.730      12.70          949   

                 dominant_rock_type  \
0                           Foidite   
1             Basalt / Picro-Basalt   
2  Trachybasalt / Tephrite Basanite   
3             Basalt / Picro-Basalt   
4                           Foidite   

                               tectonic_setting  
0        Rift zone / Continental crust (>25 km)  
1        Rift zone / Continental crust (>25 km)  
2       Intraplate / Continental crust (>25 km)  
3       Intraplate / Continental crust (>25 km)  
4  Subduction zone / Continental crust (>25 km)

Ahora, es conveniente llenar los valores faltantes de los campos `activity_evidence`, `dominant_rock_type` y `tectonic_setting` que no son determinantes para nuestro análisis.

Dado el tipo de campo y la información que contine estos valores nulos, pueden ser llenados con la cadena **No define** sin que representen algún problema para el análisis de dichos campos.

In [ ]:
df['activity_evidence'] = df['activity_evidence'].fillna('No define') 
df['dominant_rock_type'] = df['dominant_rock_type'].fillna('No define')
df['tectonic_setting'] = df['tectonic_setting'].fillna('No define')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1332 entries, 0 to 1331
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   volcano_name          1332 non-null   object 
 1   country               1332 non-null   object 
 2   primary_volcano_type  1332 non-null   object 
 3   activity_evidence     1332 non-null   object 
 4   last_known_eruption   1332 non-null   object 
 5   region                1332 non-null   object 
 6   subregion             1332 non-null   object 
 7   latitude              1332 non-null   float64
 8   longitude             1332 non-null   float64
 9   elevation_m           1332 non-null   int64  
 10  dominant_rock_type    1332 non-null   object 
 11  tectonic_setting      1332 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 125.0+ KB


##Eliminación de NaNs

Se eliminan de forma precautoria las filas y columnas donde todos los valores sean NaNs.

In [ ]:
df = df.dropna(axis=0, how='all')
df= df.dropna(axis=1, how='all')

Y validamos.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1332 entries, 0 to 1331
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   volcano_name          1332 non-null   object 
 1   country               1332 non-null   object 
 2   primary_volcano_type  1332 non-null   object 
 3   activity_evidence     1332 non-null   object 
 4   last_known_eruption   1332 non-null   object 
 5   region                1332 non-null   object 
 6   subregion             1332 non-null   object 
 7   latitude              1332 non-null   float64
 8   longitude             1332 non-null   float64
 9   elevation_m           1332 non-null   int64  
 10  dominant_rock_type    1332 non-null   object 
 11  tectonic_setting      1332 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 135.3+ KB


No se observa que se hayan eliminado líneas o columnas completos.

Se termina esta limpieza preliminar.

##Agregaciones

En un primer análisis de los datos podemos obtener alguna información relacionada con promedio, máximos y mínimos de elevación de los volcanes.

In [ ]:
## Elevación Promedio.
print("Elevación promedio: \n")
df['elevation_m'].mean()

Elevación promedio: 



1690.6103603603603

In [ ]:
## Elevación Máxima.
print('Elevación máxima: \n')
df['elevation_m'].max()

Elevación máxima: 



6879

Que corresponde a:

In [ ]:
i = df['elevation_m'].max() 
df.loc[df['elevation_m'] == i]


volcano_name          country primary_volcano_type  \
1147  Ojos del Salado, Nevados  Chile-Argentina        Stratovolcano   

     activity_evidence last_known_eruption         region  \
1147    Eruption Dated              750 CE  South America   

                                  subregion  latitude  longitude  elevation_m  \
1147  Northern Chile, Bolivia and Argentina   -27.109    -68.541         6879   

     dominant_rock_type                              tectonic_setting  
1147             Dacite  Subduction zone / Continental crust (>25 km)

In [ ]:
## Elevación Mínima.
print('Elevación mínima: \n')
df['elevation_m'].min()

Elevación mínima: 



-5700

En este otro caso corresponde a:

In [ ]:
n = df['elevation_m'].min()
df.loc[df['elevation_m']==n]

volcano_name            country primary_volcano_type  \
937  Udintsev Transform  Undersea Features            Submarine   

      activity_evidence last_known_eruption                    region  \
937  Evidence Uncertain             Unknown  Hawaii and Pacific Ocean   

                    subregion  latitude  longitude  elevation_m  \
937  Pacific Ocean (southern)   -56.153   -143.373        -5700   

    dominant_rock_type                      tectonic_setting  
937  No Data (checked)  Intraplate / Oceanic crust (< 15 km)

Este último dato nos indica que la altura se está midiendo con respecto al nivel del mar. Por lo que para encontrar una elevación media y mínima más precisas descartaremos aquellas elevaciones negativas dado que en realidad no se tiene el dato de la altura de la formación volcánica para este tipo de volcán desde su base hasta la parte superior del cono. Para los demás volcanes la base de su altura sí es el nivel medio del mar.

In [ ]:
print('Altura media sobre el nivel del mar: \n')
df['elevation_m'][df['elevation_m']>=0].mean()

Altura media sobre el nivel del mar: 



1932.6764946764947

In [ ]:
print('Altura máxima sobre el nivel del mar: \n')
df['elevation_m'][df['elevation_m']>=0].max()

Altura máxima sobre el nivel del mar: 



6879

In [ ]:
print('Altura mínima sobre el nivel del mar: \n')
df['elevation_m'][df['elevation_m']>=0].min()

Altura mínima sobre el nivel del mar: 



0

##Agrupaciones

Procedemos a identificar todos los tipos de volcanes existentes.

In [ ]:
df.groupby('primary_volcano_type')['primary_volcano_type'].count().reset_index(name='count')

primary_volcano_type  count
0               Caldera     72
1            Caldera(s)      8
2               Complex     52
3           Complex(es)      1
4              Compound     13
5               Cone(s)      1
6           Crater rows      6
7   Explosion crater(s)      1
8          Fissure vent      4
9       Fissure vent(s)     15
10            Lava cone      2
11        Lava cone(es)      1
12         Lava cone(s)      1
13            Lava dome      8
14         Lava dome(s)     29
15                 Maar      5
16              Maar(s)     10
17     Pyroclastic cone     12
18  Pyroclastic cone(s)     84
19   Pyroclastic shield      8
20               Shield    102
21            Shield(s)     22
22        Stratovolcano    548
23    Stratovolcano(es)    114
24       Stratovolcano?      2
25           Subglacial      5
26            Submarine    114
27        Submarine(es)      2
28            Tuff cone      3
29         Tuff cone(s)      9
30         Tuff ring(s)      1
31       Volcanic field     77

Vemos que en principio hay 32 tipos de volcanes, sin embargo, podemos notar que algunos de ellos son solamente el plural de los otros valores o corresponden a errores tipográficos, por lo que podemos hacer otra limpieza de los datos para obtener los valores únicos que corresponden realmente.

In [ ]:
#Creamos la función que limpia los valores de 'primary_volcano_type'
def clean_values(x):
  x = re.sub('\(\w*\)|\W','',x)
  return x
  
#Aplicamos a la columna
df['primary_volcano_type'] = df['primary_volcano_type'].map(lambda x: clean_values(x))


Y agrupamos nuevamente.

In [ ]:
volcano_types_df = df.groupby('primary_volcano_type')['primary_volcano_type'].count().reset_index(name='count').sort_values('count', ascending=False).reset_index(drop=True)
volcano_types_df['pct'] = (volcano_types_df['count']/volcano_types_df['count'].sum())*100 
volcano_types_df

primary_volcano_type  count        pct
0         Stratovolcano    664  49.849850
1                Shield    124   9.309309
2             Submarine    116   8.708709
3       Pyroclasticcone     96   7.207207
4               Caldera     80   6.006006
5         Volcanicfield     77   5.780781
6               Complex     53   3.978979
7              Lavadome     37   2.777778
8           Fissurevent     19   1.426426
9                  Maar     15   1.126126
10             Compound     13   0.975976
11             Tuffcone     12   0.900901
12    Pyroclasticshield      8   0.600601
13           Craterrows      6   0.450450
14           Subglacial      5   0.375375
15             Lavacone      4   0.300300
16      Explosioncrater      1   0.075075
17                 Cone      1   0.075075
18             Tuffring      1   0.075075

Vemos que en realidad hay 19 tipos de volcanes.
> <font color='crimson'>**Responde al plateamiento: 4. ¿Qué tipos de volcanes existen?**</font>

###Clasificación de Volcanes

In [ ]:
categorias_porc = volcano_types_df['pct'][:7].sum()
print(f'Porcentaje de categorías: {int(categorias_porc)}%')

Porcentaje de categorías: 90%


Notamos que las primeras 7 categorías agrupan a más del 90% de los volcanes

###Uso de Merge

Y ahora, nos proponemos a obtener la altura media por tipo de volcán.

In [ ]:
avg_elevation = df.groupby('primary_volcano_type')['elevation_m'].mean().reset_index(name='avg_elevation')
volcano_types_df = pd.merge(volcano_types_df,avg_elevation, left_on='primary_volcano_type', right_on='primary_volcano_type')
volcano_types_df[['primary_volcano_type','avg_elevation']]

primary_volcano_type  avg_elevation
0         Stratovolcano    2203.597892
1                Shield    1509.137097
2             Submarine    -909.637931
3       Pyroclasticcone    1660.760417
4               Caldera    1475.450000
5         Volcanicfield    1572.402597
6               Complex    2290.603774
7              Lavadome    1646.702703
8           Fissurevent     973.473684
9                  Maar    1238.733333
10             Compound    2600.000000
11             Tuffcone    1221.500000
12    Pyroclasticshield    2413.000000
13           Craterrows     526.500000
14           Subglacial    1518.000000
15             Lavacone    2129.750000
16      Explosioncrater     -48.000000
17                 Cone    1326.000000
18             Tuffring     830.000000

### Volcanes por ubicación geográfica

Nuestro siguiente propósito es agrupar los volcanes por su ubicación geográfica para encontrar las regiones de mayor riesgo.

Empezaremos por agruparlos por región.

> <font color='crimson'>**Responde al plateamiento: 6. ¿Qué país/región tiene la mayor cantidad de volcanes?**</font>


In [ ]:
volcanos_per_region = df.groupby('region')['region'].count().reset_index(name='count').sort_values('count', ascending=False).reset_index(drop=True)
volcanos_per_region

region  count
0                    South America    176
1          Japan, Taiwan, Marianas    134
2                        Indonesia    123
3               Africa and Red Sea    119
4       México and Central America    109
5      Kamchatka and Mainland Asia     86
6                           Alaska     83
7          Melanesia and Australia     73
8           Canada and Western USA     64
9              New Zealand to Fiji     55
10                   Kuril Islands     44
11         Philippines and SE Asia     43
12    Middle East and Indian Ocean     41
13  Mediterranean and Western Asia     40
14        Iceland and Arctic Ocean     37
15        Hawaii and Pacific Ocean     33
16                  Atlantic Ocean     28
17                      Antarctica     28
18                     West Indies     16

Vemos que la región con mayor riesgo es **South America**

Ahora por subregión.

In [ ]:
volcanos_per_subregion = df.groupby('subregion')['subregion'].count().reset_index(name='count').sort_values('count', ascending=False).reset_index(drop=True)
volcanos_per_subregion.head(30)

subregion  count
0                     Kamchatka Peninsula     67
1       Africa (northeastern) and Red Sea     63
2       Izu, Volcano, and Mariana Islands     51
3   Northern Chile, Bolivia and Argentina     46
4                                  Honshu     45
5                           Kuril Islands     44
6             Central Chile and Argentina     39
7                        Aleutian Islands     37
8                                  Mexico     37
9                                    Java     36
10                       Alaska Peninsula     30
11  Antarctica and South Sandwich Islands     28
12                       Africa (eastern)     27
13                                Sumatra     27
14           Southern Chile and Argentina     26
15                   Lesser Sunda Islands     23
16               El Salvador and Honduras     22
17                                Ecuador     21
18                                 Canada     21
19                              Guatemala     21
20                          Tonga Islands     21
21              Ryukyu Islands and Kyushu     18
22                               Hokkaido     18
23                            New Zealand     17
24                              Nicaragua     17
25                            New Britain     16
26                              Halmahera     16
27                                   Peru     16
28                            West Indies     16
29                                  Luzon     15

En este caso vemos que las subregiones con mayor riesgo son prácticamente 3:
- **Kamchatka Peninsula**, 
- **Africa (northeastern) and Red Sea**
- **Izu, Volcano, and Mariana Islands**

Y por, último, por cada país.
> <font color='crimson'>**Responde al plateamiento: 6. ¿Qué país/continente/región tiene la mayor cantidad de volcanes?**</font>

In [ ]:
volcanos_per_country = df.groupby('country')['country'].count().reset_index(name='count').sort_values('count', ascending=False).reset_index(drop=True)
volcanos_per_country.head(30)

country  count
0                    United States    162
1                        Indonesia    121
2                            Japan    107
3                           Russia    102
4                            Chile     66
5                         Ethiopia     49
6                 Papua New Guinea     46
7                      Philippines     38
8                           Mexico     36
9                          Iceland     35
10                         Ecuador     34
11                     New Zealand     24
12               Undersea Features     24
13                          Canada     24
14                           Tonga     21
15                           Kenya     21
16                       Guatemala     20
17                          France     19
18                      Antarctica     19
19                     El Salvador     18
20                 Chile-Argentina     18
21                       Nicaragua     17
22                       Argentina     17
23                            Peru     16
24  Japan - administered by Russia     15
25                         Vanuatu     14
26                           Italy     13
27                        Colombia     13
28                  United Kingdom     13
29                        Portugal     13

Al nivel de país es **United States** quien tiene el mayor riesgo por su densidad volcánica.

México aparece en la 9° posición.

### Relación Tipo de Volcán vs Tectónica de Placas.

Ahora queremos ver la relación entre el tipo de volcán y la tectónica de placas. Empecemos viendo los valores para `tectonic_setting`.
><font color='crimson'>**Responde al plateamiento: 8. Relaciones tectónica de placas - tipo de volcán.**</font>

In [ ]:
df['tectonic_setting'].unique()

array(['Rift zone / Continental crust (>25 km)',
       'Intraplate / Continental crust (>25 km)',
       'Subduction zone / Continental crust (>25 km)', 'No define',
       'Rift zone / Oceanic crust (< 15 km)',
       'Rift zone / Intermediate crust (15-25 km)',
       'Intraplate / Oceanic crust (< 15 km)',
       'Subduction zone / Oceanic crust (< 15 km)',
       'Subduction zone / Crustal thickness unknown',
       'Subduction zone / Intermediate crust (15-25 km)', 'Unknown',
       'Intraplate / Intermediate crust (15-25 km)'], dtype=object)

Vemos que los valores `No define` y `Unkown` no nos sirven, por lo que los excluimos del DataFrame.

In [ ]:
tectonic_setting = df[~df['tectonic_setting'].isin(['Unknown','No define'])]
tectonic_setting[['tectonic_setting','primary_volcano_type']].groupby(['tectonic_setting','primary_volcano_type']).size().head(40)

tectonic_setting                            primary_volcano_type
Intraplate / Continental crust (>25 km)     Caldera                  2
                                            Fissurevent              1
                                            Lavacone                 1
                                            Lavadome                 1
                                            Maar                     1
                                            Pyroclasticcone         39
                                            Pyroclasticshield        1
                                            Shield                  22
                                            Stratovolcano           31
                                            Subglacial               1
                                            Submarine                2
                                            Tuffcone                 1
                                            Volcanicfield           31
Intraplate /

#<font size=6 color='dodgerblue'>6. Transformación, filtrado y ordenamiento de datos.</font>

---



##Procesamiento de campo `last_known_eruption`

###División del campo.

Respecto a la fecha de la última erupción conocida `last_known_eruption`se ve la conveniencia de separar la columna en sus componentes de Año y Era.
Esto debido a que la fecha está representada por la línea de tiempo bajo la siguiente nomenclatura:
- **BCE**: Before Common Era or Before Current Era or Before Christian Era: used when referring to a year before the birth of Jesus Christ when the Christian calendar starts counting years.
- **CE**: Common Era or Christian era: used when referring to a year after the birth of Jesus Christ when the Christian calendar starts counting years.

In [ ]:
df_last_known_eruption = df.copy()
df_last_known_eruption['last_known_eruption'].str.split('\s', expand=True)

0     1
0        8300   BCE
1        4040   BCE
2     Unknown  None
3        3600   BCE
4     Unknown  None
...       ...   ...
1327  Unknown  None
1328     2016    CE
1329     1962    CE
1330  Unknown  None
1331  Unknown  None

[1332 rows x 2 columns]

Se adicionan nuevas columnas `last_known_eruption_year` y `last_known_eruption_era` al DataFrame.

In [ ]:
df_last_known_eruption[['last_known_eruption_year', 'last_known_eruption_era']] = df['last_known_eruption'].str.split('\s', expand=True)

Validamos los nuevos campos.

In [ ]:
df_last_known_eruption[['last_known_eruption', 'last_known_eruption_year', 'last_known_eruption_era']].head(2)

last_known_eruption last_known_eruption_year last_known_eruption_era
0            8300 BCE                     8300                     BCE
1            4040 BCE                     4040                     BCE

In [ ]:
df_last_known_eruption[['last_known_eruption', 'last_known_eruption_year', 'last_known_eruption_era']].dtypes

last_known_eruption         object
last_known_eruption_year    object
last_known_eruption_era     object
dtype: object

Como se observa es necesario cambiar el tipo de dato del campo `last_known_eruption_year`, el cual se presenta como object y debe ser del tipo int64.



Empleamos el método **to_numeric** con la opción **coerce** para forzar la conversión.

In [ ]:
pd.to_numeric(df_last_known_eruption['last_known_eruption_year'], errors='coerce')

0       8300.0
1       4040.0
2          NaN
3       3600.0
4          NaN
         ...  
1327       NaN
1328    2016.0
1329    1962.0
1330       NaN
1331       NaN
Name: last_known_eruption_year, Length: 1332, dtype: float64

Ahora detectamos que tenemos **NaNs** en los registros de la nueva columna, como este dato es muy importante para nuestro proyecto, y no hay forma de obtener el valor de este campo para los registros convertidos, debemos de eliminar tales registros.

Antes reasignamos la columna al DataSet.

In [ ]:
df_last_known_eruption['last_known_eruption_year'] = pd.to_numeric(df_last_known_eruption['last_known_eruption_year'], errors='coerce')
df_last_known_eruption['last_known_eruption_year']

0       8300.0
1       4040.0
2          NaN
3       3600.0
4          NaN
         ...  
1327       NaN
1328    2016.0
1329    1962.0
1330       NaN
1331       NaN
Name: last_known_eruption_year, Length: 1332, dtype: float64

### Eliminación de NaNs

Para poder continuar con la conversión del campo al tipo int64 es necesaria la eliminación de los registros con valores NaNs, en este caso por renglón, ya que de no realizar la eliminación se presentarán errores en la conversión a int64.

In [ ]:
df_no_nans = df.copy()
df_no_nans['activity_evidence'] = df['activity_evidence'].fillna('No define') 
df_no_nans['dominant_rock_type'] = df_no_nans['dominant_rock_type'].fillna('No define')
df_no_nans['tectonic_setting'] = df_no_nans['tectonic_setting'].fillna('No define')
df_no_nans.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1332 entries, 0 to 1331
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   volcano_name          1332 non-null   object 
 1   country               1332 non-null   object 
 2   primary_volcano_type  1332 non-null   object 
 3   activity_evidence     1332 non-null   object 
 4   last_known_eruption   1332 non-null   object 
 5   region                1332 non-null   object 
 6   subregion             1332 non-null   object 
 7   latitude              1332 non-null   float64
 8   longitude             1332 non-null   float64
 9   elevation_m           1332 non-null   int64  
 10  dominant_rock_type    1332 non-null   object 
 11  tectonic_setting      1332 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 135.3+ KB


Asignamos el DataFrame sin NaNs a nuestro DF de trabajo.

In [ ]:
df = df_no_nans

¡Listo!
Se puede continuar con la eliminación de las líneas con NaNs en el campo `last_known_eruption_year`.

In [ ]:
df_last_known_eruption = df_last_known_eruption.dropna(axis=0, how='any')

Validamos la depuración usando **head** y **tail** que previamente nos mostraban los registros con valores de **Unknow**.

> <font color='crimson'>**Responde al plateamiento: 3. ¿De cuáles se tiene registro de -al menos de una- erupción?**</font>. En este caso son 1,329


In [ ]:
df_last_known_eruption.head()

volcano_name  country primary_volcano_type  activity_evidence  \
0  West Eifel Volcanic Field  Germany                 Maar     Eruption Dated   
1            Chaine des Puys   France             Lavadome     Eruption Dated   
3   Calatrava Volcanic Field    Spain      Pyroclasticcone     Eruption Dated   
5              Campi Flegrei    Italy              Caldera  Eruption Observed   
6                   Vesuvius    Italy        Stratovolcano  Eruption Observed   

  last_known_eruption                          region       subregion  \
0            8300 BCE  Mediterranean and Western Asia  Western Europe   
1            4040 BCE  Mediterranean and Western Asia  Western Europe   
3            3600 BCE  Mediterranean and Western Asia  Western Europe   
5             1538 CE  Mediterranean and Western Asia           Italy   
6             1944 CE  Mediterranean and Western Asia           Italy   

   latitude  longitude  elevation_m                  dominant_rock_type  \
0    50.170      6.850          600                             Foidite   
1    45.775      2.970         1464               Basalt / Picro-Basalt   
3    38.870     -4.020         1117               Basalt / Picro-Basalt   
5    40.827     14.139          458             Trachyte / Trachydacite   
6    40.821     14.426         1281  Phono-tephrite /  Tephri-phonolite   

                               tectonic_setting  last_known_eruption_year  \
0        Rift zone / Continental crust (>25 km)                    8300.0   
1        Rift zone / Continental crust (>25 km)                    4040.0   
3       Intraplate / Continental crust (>25 km)                    3600.0   
5  Subduction zone / Continental crust (>25 km)                    1538.0   
6  Subduction zone / Continental crust (>25 km)                    1944.0   

  last_known_eruption_era  
0                     BCE  
1                     BCE  
3                     BCE  
5                      CE  
6                      CE

In [ ]:
df_last_known_eruption.tail()

volcano_name         country primary_volcano_type  \
1323       Montagu Island  United Kingdom               Shield   
1324             Saunders  United Kingdom        Stratovolcano   
1325     Candlemas Island  United Kingdom        Stratovolcano   
1328           Zavodovski  United Kingdom        Stratovolcano   
1329  Protector Seamounts  United Kingdom            Submarine   

      activity_evidence last_known_eruption      region  \
1323  Eruption Observed             2007 CE  Antarctica   
1324  Eruption Observed             2022 CE  Antarctica   
1325  Eruption Observed             1911 CE  Antarctica   
1328  Eruption Observed             2016 CE  Antarctica   
1329  Eruption Observed             1962 CE  Antarctica   

                                  subregion  latitude  longitude  elevation_m  \
1323  Antarctica and South Sandwich Islands   -58.445    -26.374         1370   
1324  Antarctica and South Sandwich Islands   -57.800    -26.483          843   
1325  Antarctica and South Sandwich Islands   -57.080    -26.670          550   
1328  Antarctica and South Sandwich Islands   -56.300    -27.570          551   
1329  Antarctica and South Sandwich Islands   -55.912    -28.167          -55   

         dominant_rock_type                           tectonic_setting  \
1323  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
1324  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
1325  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
1328  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
1329               Rhyolite  Subduction zone / Oceanic crust (< 15 km)   

      last_known_eruption_year last_known_eruption_era  
1323                    2007.0                      CE  
1324                    2022.0                      CE  
1325                    1911.0                      CE  
1328                    2016.0                      CE  
1329                    1962.0                      CE

Vemos nuevamente el detalle de las columnas.

In [ ]:
df_last_known_eruption.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 860 entries, 0 to 1329
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   volcano_name              860 non-null    object 
 1   country                   860 non-null    object 
 2   primary_volcano_type      860 non-null    object 
 3   activity_evidence         860 non-null    object 
 4   last_known_eruption       860 non-null    object 
 5   region                    860 non-null    object 
 6   subregion                 860 non-null    object 
 7   latitude                  860 non-null    float64
 8   longitude                 860 non-null    float64
 9   elevation_m               860 non-null    int64  
 10  dominant_rock_type        860 non-null    object 
 11  tectonic_setting          860 non-null    object 
 12  last_known_eruption_year  860 non-null    float64
 13  last_known_eruption_era   860 non-null    object 
dtypes: float6

La cantidad de registros que se conservan en nuestro DataFrame son 860 de los 1332 originales y en este momento tenemos 14 columnas, aumentando 1 a las 13 originales cuando se normalizó el campo `Volcano Number`.


> Se eliminó la columna de `Volcano Number` pero se adicionaron 2 columnas al dividir `last_know_eruption` en sus 2 componentes de año y era.



###Reindexado

Realizamos la reindexación de nuestro DataFrame solo a manera de validar que el número de filas corresponde al índice.

In [ ]:
df_last_known_eruption.reset_index(drop=True).tail()

volcano_name         country primary_volcano_type  \
855       Montagu Island  United Kingdom               Shield   
856             Saunders  United Kingdom        Stratovolcano   
857     Candlemas Island  United Kingdom        Stratovolcano   
858           Zavodovski  United Kingdom        Stratovolcano   
859  Protector Seamounts  United Kingdom            Submarine   

     activity_evidence last_known_eruption      region  \
855  Eruption Observed             2007 CE  Antarctica   
856  Eruption Observed             2022 CE  Antarctica   
857  Eruption Observed             1911 CE  Antarctica   
858  Eruption Observed             2016 CE  Antarctica   
859  Eruption Observed             1962 CE  Antarctica   

                                 subregion  latitude  longitude  elevation_m  \
855  Antarctica and South Sandwich Islands   -58.445    -26.374         1370   
856  Antarctica and South Sandwich Islands   -57.800    -26.483          843   
857  Antarctica and South Sandwich Islands   -57.080    -26.670          550   
858  Antarctica and South Sandwich Islands   -56.300    -27.570          551   
859  Antarctica and South Sandwich Islands   -55.912    -28.167          -55   

        dominant_rock_type                           tectonic_setting  \
855  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
856  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
857  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
858  Basalt / Picro-Basalt  Subduction zone / Oceanic crust (< 15 km)   
859               Rhyolite  Subduction zone / Oceanic crust (< 15 km)   

     last_known_eruption_year last_known_eruption_era  
855                    2007.0                      CE  
856                    2022.0                      CE  
857                    1911.0                      CE  
858                    2016.0                      CE  
859                    1962.0                      CE

In [ ]:
df_last_known_eruption.shape

(860, 14)

###Conversión de `last_known_eruption_year` a int64

Terminamos la conversión, ya que con los pasos anteriores, el campo ha quedado definido como float64 como podemos ver a continuación.

In [ ]:
df_last_known_eruption['last_known_eruption_year'].dtypes

dtype('float64')

In [ ]:
## Para este fin nos auxiliaremos de un pequeño diccionario.
diccionario_year = {
    'last_known_eruption_year': int
}

In [ ]:
## Pasamos el diccionario al método astype del DataFrame:
df_last_known_eruption.astype(diccionario_year).dtypes

volcano_name                 object
country                      object
primary_volcano_type         object
activity_evidence            object
last_known_eruption          object
region                       object
subregion                    object
latitude                    float64
longitude                   float64
elevation_m                   int64
dominant_rock_type           object
tectonic_setting             object
last_known_eruption_year      int64
last_known_eruption_era      object
dtype: object

La conversión ha sido exitosa.

<font size=6 color='dodgerblue'> 7. Visualización de los datos: Mapas interactivos


---



Ahora que ya hemos hecho una limpieza y análisis de datos, dado el tipo de información con la que hemos trabajado, creemos conveniente visualizarla a través de mapas.
Para ello nos apoyaremos de la biblioteca '*folium*'.

## Mapa 1: Volcanes del mundo

En este primer mapa ubicaremos todos los volcanes contenidos en el dataset original.
> <font color='crimson'>**Responde al plateamiento: 5. ¿Dónde se ubican?**</font>

Lo primero que realizamos es crear nuestro mapa base.

In [ ]:
#!pip install folium

import folium
world_map = folium.Map(location=[15,-20], zoom_start=2.3, tiles= 'Stamen Terrain')

Luego, creamos un objeto `MarkerCluster` y lo anexamos al mapa, este objeto nos permitirá agrupar los volcanes cercanos entre sí lo que nos ayudará a ubicarlos por regiones.

In [ ]:
from folium.plugins import MarkerCluster
marker_cluster = MarkerCluster()
world_map.add_child(marker_cluster)

Por último, agregamos un marcador para cada volcán, lo agrupamos en el `MarkerCluster` y automáticamente se agrega al mapa.
Y,... ¡vualá!, tenemos nuestro primer mapa.

In [ ]:
for lat,lng,name,v_type in zip(df.latitude, df.longitude,df.volcano_name,df.primary_volcano_type):
    marker= folium.features.Marker(
            location= [lat,lng],
            tooltip=name,
            popup=v_type,
            icon= folium.features.CustomIcon(
                icon_image='https://volcano.si.edu/includes/images/volcano_icon_24.png',
                    icon_size=(20,20))
            )
    marker_cluster.add_child(marker)
world_map

## Mapa 2: Volcanes de México

En este mapa, nos proponemos a visualizar los volcanes que se encuentran exclusivamente en territorio nacional, y sus zonas de riesgo alto (10 km) y medio (30 km).

La construcción del mapa es similar al anterior, pero tiene sus sutiles diferencias.

Primero, a partir de dataset principal creamos el subconjunto con el que se trabajó.

In [ ]:
volcanes_de_mexico = df[df.country=='Mexico']
volcanes_de_mexico

volcano_name country primary_volcano_type  \
938             Prieto, Cerro  Mexico             Lavadome   
939                  Pinacate  Mexico      Pyroclasticcone   
940            San Luis, Isla  Mexico             Tuffcone   
941   Jaraguay Volcanic Field  Mexico      Pyroclasticcone   
942                  Coronado  Mexico        Stratovolcano   
943                 Guadalupe  Mexico               Shield   
944  San Borja Volcanic Field  Mexico      Pyroclasticcone   
945             Tortuga, Isla  Mexico               Shield   
946       Comondu-La Purisima  Mexico        Volcanicfield   
947                   Barcena  Mexico      Pyroclasticcone   
948                   Socorro  Mexico               Shield   
949    Durango Volcanic Field  Mexico      Pyroclasticcone   
950              Isabel, Isla  Mexico             Tuffcone   
951                Sanganguey  Mexico        Stratovolcano   
952                  Ceboruco  Mexico        Stratovolcano   
953    Mascota Volcanic Field  Mexico      Pyroclasticcone   
954                    Colima  Mexico        Stratovolcano   
955      Michoacan-Guanajuato  Mexico      Pyroclasticcone   
956  Zitacuaro-Valle de Bravo  Mexico              Caldera   
957                Jocotitlan  Mexico        Stratovolcano   
958         Toluca, Nevado de  Mexico        Stratovolcano   
959             Chichinautzin  Mexico        Volcanicfield   
960                    Papayo  Mexico             Lavadome   
961              Iztaccihuatl  Mexico        Stratovolcano   
962              Popocatepetl  Mexico        Stratovolcano   
963              Malinche, La  Mexico        Stratovolcano   
964           Serdan-Oriental  Mexico             Tuffcone   
965              Humeros, Los  Mexico              Caldera   
966             Atlixcos, Los  Mexico               Shield   
967   Naolinco Volcanic Field  Mexico      Pyroclasticcone   
968           Cofre de Perote  Mexico             Compound   
969                Gloria, La  Mexico        Volcanicfield   
970             Cumbres,  Las  Mexico        Stratovolcano   
971          Orizaba, Pico de  Mexico        Stratovolcano   
972                San Martin  Mexico               Shield   
973               Chichon, El  Mexico             Lavadome   

      activity_evidence last_known_eruption                      region  \
938  Evidence Uncertain             Unknown  México and Central America   
939   Evidence Credible             Unknown  México and Central America   
940   Evidence Credible             Unknown  México and Central America   
941   Evidence Credible             Unknown  México and Central America   
942   Unrest / Holocene             Unknown  México and Central America   
943   Evidence Credible             Unknown  México and Central America   
944   Evidence Credible             Unknown  México and Central America   
945   Evidence Credible             Unknown  México and Central America   
946  Evidence Uncertain             Unknown  México and Central America   
947   Eruption Observed             1953 CE  México and Central America   
948   Eruption Observed             1994 CE  México and Central America   
949   Evidence Credible             Unknown  México and Central America   
950  Evidence Uncertain             Unknown  México and Central America   
951   Evidence Credible             Unknown  México and Central America   
952   Eruption Observed             1875 CE  México and Central America   
953   Evidence Credible             Unknown  México and Central America   
954   Eruption Observed             2019 CE  México and Central America   
955   Eruption Observed             1952 CE  México and Central America   
956      Eruption Dated            3050 BCE  México and Central America   
957      Eruption Dated             1270 CE  México and Central America   
958      Eruption Dated            1350 BCE  México and Central America   
959      Eruption Dated              400 CE  México and Central America 

Creamos el mapa base y el objeto que agrupará todos los marcadores.

In [ ]:
mexico_map = folium.Map(location=[22,-100],zoom_start=5.4, tiles='Stamen Terrain')
ubicaciones = folium.map.FeatureGroup()

Ahora, agregamos un identificador para cada volcán y dos círculos representando las zonas de riesgo alto (rojo) y medio (naranja).

In [ ]:
# Iteramos dentro del dataset
for lat,lng,name,v_type in zip(volcanes_de_mexico.latitude, volcanes_de_mexico.longitude,volcanes_de_mexico.volcano_name,volcanes_de_mexico.primary_volcano_type):
    ubicaciones.add_child(           # Primer objeto: identificador del volcán
        folium.features.Marker(              
            location= [lat,lng],
            tooltip=name,
            popup=v_type,
            icon= folium.features.CustomIcon(
                icon_image='https://volcano.si.edu/includes/images/volcano_icon_24.png',
                    icon_size=(20,20))
            )
     )
    ubicaciones.add_child(         # Segundo objeto: círculo interno
        folium.features.Circle(
            location=[lat,lng],
            color='red',
            fill = True,
            radius = 10000,
            popup = 'Zona de alto riesgo: 10 km'
            )
    )
    ubicaciones.add_child(         # Tercer objeto: círculo externo
        folium.features.Circle(
            location=[lat,lng],
            color='orange',
            fill = True,
            radius = 30000,
            popup = 'Zona de riesgo medio: 30 km'
            )
    )

Agregamos al mapa y visualizamos.

In [ ]:
mexico_map.add_child(ubicaciones)
mexico_map

## Mapa 3: Volcanes de alto riesgo

Este tercer mapa pretende ubicar a los volcanes que mayor peligro presentan para la población de acuerdo a su historial de erupciones destructivas frecuentes y su cercanía a zonas densamente pobladas.

><font color='crimson'>**Responde al plateamiento: 1. ¿Cuáles volcanes representan mayor riesgo?**</font>

###web scrapping
Para este mapa se requirió de técnicas de 'web scrapping' ya que la información no se encontraba disponible en el data set original y se obtuvo de [Wikipedia - Decade Volcanoes](https://en.wikipedia.org/wiki/Decade_Volcanoes).

Empezamos importando las bibliotecas requeridas para el 'web scraping'

In [ ]:
#!pip install bs4
import requests 
from bs4 import BeautifulSoup
import re

Obtenemos la liga, hacemos la solicitud y revisamos el estado de ésta.

In [ ]:
url = 'https://en.wikipedia.org/wiki/Decade_Volcanoes'
req = requests.get(url)
req.status_code

200

Recopilamos la información necesaria y creamos un `BS object` que nos permitirá extraer información del archivo html.

In [ ]:
data = req.text
bs_object = BeautifulSoup(data,'lxml')

Ubicamos la tabla dentro del objeto y creamos un DataFrame.

In [ ]:
table = bs_object.find('table')
decade_volcanos=pd.read_html(str(table), flavor='lxml')[0]
decade_volcanos

Volcano                                          Elevation  \
0   Avachinsky-Koryaksky  2,741 m (8,993 ft) (Avachinsky)3,456 m (11,339...   
1       Volcán de Colima                                3,820 m (12,533 ft)   
2                Galeras                                4,276 m (14,029 ft)   
3              Mauna Loa                                4,169 m (13,678 ft)   
4             Mount Etna                                3,326 m (10,912 ft)   
5           Mount Merapi                                 2,910 m (9,547 ft)   
6       Mount Nyiragongo                                3,470 m (11,385 ft)   
7          Mount Rainier                                4,392 m (14,409 ft)   
8         Mount Vesuvius                                 1,281 m (4,203 ft)   
9            Mount Unzen                                 1,486 m (4,875 ft)   
10            Sakurajima                                 1,117 m (3,665 ft)   
11           Santa María                                3,772 m (12,375 ft)   
12     Santorini (Thera)                                   367 m (1,204 ft)   
13          Taal Volcano                                   311 m (1,020 ft)   
14                 Teide                                3,715 m (12,188 ft)   
15                Ulawun                                 2,334 m (7,657 ft)   

                                       Region  \
0                                   Kamchatka   
1                              Colima/Jalisco   
2                                      Nariño   
3                                      Hawaii   
4                                      Sicily   
5   Central Java/Special Region of Yogyakarta   
6                                  North Kivu   
7                                  Washington   
8                                    Campania   
9                           Nagasaki/Kumamoto   
10                                  Kagoshima   
11                             Quetzaltenango   
12                               South Aegean   
13                                 Calabarzon   
14                             Canary Islands   
15          East New Britain/West New Britain   

                             Country               Population**  \
0                             Russia     179,367 (2021 est) [6]   
1                             Mexico    770,000 (2019 est) [10]   
2                           Colombia   1,630,000 (2018 est)[12]   
3                      United States                          ?   
4                              Italy              1,100,000[14]   
5                          Indonesia   5,050,000 (2019 est)[16]   
6   Democratic Republic of the Congo                    2000000   
7                      United States             795,000 (2010)   
8                              Italy              3,085,000[14]   
9                              Japan   1,320,000 (2019 est)[19]   
10                             Japan   1,470,000 (2019 est)[19]   
11                         Guatemala   1,127,000 (2018 est)[21]   
12                            Greece                          ?   
13                       Philippines  24,393,000 (2020 est)[24]   
14                             Spain                     900000   
15                  Papua New Guinea                          ?   

                          Last known eruption  \
0   2001 (Avachinsky),[7] 2009 (Koryaksky)[8]   
1                                    2019[11]   
2                                    2014[13]   
3                                        1984   
4                                2013-present   
5                                        2022   
6               22 May 2021-present, VEI 0-2.   
7                 1894-1895, (minor phreatic)   
8                    17-23 March, 1944, VEI 3   
9                                        1995   
10                                   2022[20]   
11                             23 August 2013   
12                                       1950   
13           

In [ ]:
decade_volcanos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 7 columns):
 #   Column                                Non-Null Count  Dtype 
---  ------                                --------------  ----- 
 0   Volcano                               16 non-null     object
 1   Elevation                             16 non-null     object
 2   Region                                16 non-null     object
 3   Country                               16 non-null     object
 4   Population**                          16 non-null     object
 5   Last known eruption                   16 non-null     object
 6   Affected Region/Why a Decade Volcano  16 non-null     object
dtypes: object(7)
memory usage: 1.0+ KB


Vemos que este DataFrame contiene mucha información que no requerimos y sólo nos estorba, por lo que aplicaremos una limpieza de datos.

Empezamos por eliminar la última columna.

In [ ]:
decade_volcanos = decade_volcanos.drop('Affected Region/Why a Decade Volcano', axis=1)
decade_volcanos

Volcano                                          Elevation  \
0   Avachinsky-Koryaksky  2,741 m (8,993 ft) (Avachinsky)3,456 m (11,339...   
1       Volcán de Colima                                3,820 m (12,533 ft)   
2                Galeras                                4,276 m (14,029 ft)   
3              Mauna Loa                                4,169 m (13,678 ft)   
4             Mount Etna                                3,326 m (10,912 ft)   
5           Mount Merapi                                 2,910 m (9,547 ft)   
6       Mount Nyiragongo                                3,470 m (11,385 ft)   
7          Mount Rainier                                4,392 m (14,409 ft)   
8         Mount Vesuvius                                 1,281 m (4,203 ft)   
9            Mount Unzen                                 1,486 m (4,875 ft)   
10            Sakurajima                                 1,117 m (3,665 ft)   
11           Santa María                                3,772 m (12,375 ft)   
12     Santorini (Thera)                                   367 m (1,204 ft)   
13          Taal Volcano                                   311 m (1,020 ft)   
14                 Teide                                3,715 m (12,188 ft)   
15                Ulawun                                 2,334 m (7,657 ft)   

                                       Region  \
0                                   Kamchatka   
1                              Colima/Jalisco   
2                                      Nariño   
3                                      Hawaii   
4                                      Sicily   
5   Central Java/Special Region of Yogyakarta   
6                                  North Kivu   
7                                  Washington   
8                                    Campania   
9                           Nagasaki/Kumamoto   
10                                  Kagoshima   
11                             Quetzaltenango   
12                               South Aegean   
13                                 Calabarzon   
14                             Canary Islands   
15          East New Britain/West New Britain   

                             Country               Population**  \
0                             Russia     179,367 (2021 est) [6]   
1                             Mexico    770,000 (2019 est) [10]   
2                           Colombia   1,630,000 (2018 est)[12]   
3                      United States                          ?   
4                              Italy              1,100,000[14]   
5                          Indonesia   5,050,000 (2019 est)[16]   
6   Democratic Republic of the Congo                    2000000   
7                      United States             795,000 (2010)   
8                              Italy              3,085,000[14]   
9                              Japan   1,320,000 (2019 est)[19]   
10                             Japan   1,470,000 (2019 est)[19]   
11                         Guatemala   1,127,000 (2018 est)[21]   
12                            Greece                          ?   
13                       Philippines  24,393,000 (2020 est)[24]   
14                             Spain                     900000   
15                  Papua New Guinea                          ?   

                          Last known eruption  
0   2001 (Avachinsky),[7] 2009 (Koryaksky)[8]  
1                                    2019[11]  
2                                    2014[13]  
3                                        1984  
4                                2013-present  
5                                        2022  
6               22 May 2021-present, VEI 0-2.  
7                 1894-1895, (minor phreatic)  
8                    17-23 March, 1944, VEI 3  
9                                        1995  
10                                   2022[20]  
11                             23 August 2013  
12                                       1950  
13                         

Pasamos los nombres de las columnas a un formato **snake case**.

In [ ]:
cols={}
for elem in [col for col in decade_volcanos.columns]:
    col = re.sub(' ','_',elem)
    col = col.lower()
    cols[elem] = col
cols['Population**']='pop_at_risk'
cols

{'Volcano': 'volcano',
 'Elevation': 'elevation',
 'Region': 'region',
 'Country': 'country',
 'Population**': 'pop_at_risk',
 'Last known eruption': 'last_known_eruption'}

Y validamos la transformación

In [ ]:
decade_volcanos.rename(columns=cols, inplace=True)
decade_volcanos.columns

Index(['volcano', 'elevation', 'region', 'country', 'pop_at_risk',
       'last_known_eruption'],
      dtype='object')

Procedemos a eliminar la información en sistema inglés en la columna `elevation`.

In [ ]:
decade_volcanos['elevation'] = decade_volcanos['elevation'].apply(lambda x: re.sub('\(.*\)|[^0-9]','',x))
decade_volcanos['elevation']

0     2741
1     3820
2     4276
3     4169
4     3326
5     2910
6     3470
7     4392
8     1281
9     1486
10    1117
11    3772
12     367
13     311
14    3715
15    2334
Name: elevation, dtype: object

Extraemos el año de la última erupción de la columna `last_known_eruption`.

In [ ]:
decade_volcanos.last_known_eruption = decade_volcanos.last_known_eruption.str.extract(r'(\d{4})')
decade_volcanos.last_known_eruption

0     2001
1     2019
2     2014
3     1984
4     2013
5     2022
6     2021
7     1894
8     1944
9     1995
10    2022
11    2013
12    1950
13    2021
14    1909
15    2019
Name: last_known_eruption, dtype: object

Limpiamos la columna `pop_at_risk` y la convertimos a un valor numérico.

In [ ]:
decade_volcanos.pop_at_risk = decade_volcanos.pop_at_risk.str.replace(r'(\[.*\])|(\(.*\))|,','')
decade_volcanos.pop_at_risk = pd.to_numeric(decade_volcanos.pop_at_risk, errors='coerce', downcast='integer')
decade_volcanos.pop_at_risk.fillna(0, inplace=True)
decade_volcanos.pop_at_risk

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  """Entry point for launching an IPython kernel.


0       179367.0
1       770000.0
2      1630000.0
3            0.0
4      1100000.0
5      5050000.0
6      2000000.0
7       795000.0
8      3085000.0
9      1320000.0
10     1470000.0
11     1127000.0
12           0.0
13    24393000.0
14      900000.0
15           0.0
Name: pop_at_risk, dtype: float64

Vemos que este data set no cuenta con las coordenadas de los volcanes, por lo que nos apoyaremos de nuestro data set principal para obtenerlas, así que verificamos la compatibilidad entre ambos.

In [ ]:
# ubicamos los volcanes por nombre en el data set principal
df_last_known_eruption[df_last_known_eruption['volcano_name'].isin(decade_volcanos.volcano)].reset_index()

index volcano_name           country primary_volcano_type  \
0    281       Ulawun  Papua New Guinea        Stratovolcano   
1    917    Mauna Loa     United States               Shield   
2   1057      Galeras          Colombia              Complex   

   activity_evidence last_known_eruption                    region  \
0  Eruption Observed             2022 CE   Melanesia and Australia   
1  Eruption Observed             1984 CE  Hawaii and Pacific Ocean   
2  Eruption Observed             2014 CE             South America   

          subregion  latitude  longitude  elevation_m  \
0       New Britain    -5.050    151.330         2334   
1  Hawaiian Islands    19.475   -155.608         4170   
2          Colombia     1.220    -77.370         4276   

             dominant_rock_type                              tectonic_setting  \
0         Basalt / Picro-Basalt  Subduction zone / Continental crust (>25 km)   
1         Basalt / Picro-Basalt          Intraplate / Oceanic crust (< 15 km)   
2  Andesite / Basaltic Andesite  Subduction zone / Continental crust (>25 km)   

   last_known_eruption_year last_known_eruption_era  
0                    2022.0                      CE  
1                    1984.0                      CE  
2                    2014.0                      CE

Vemos que solamente tres nombres coinciden, por lo que también necesitamos modificar la columna `volcano` para que coincidan con el data set original.

Lo primero que hacemos es eliminar la palabra 'Mount' del nombre.
Y corroboramos

In [ ]:
decade_volcanos.volcano = decade_volcanos.volcano.str.replace('Mount ','')
df_last_known_eruption[df_last_known_eruption['volcano_name'].isin(decade_volcanos.volcano)].reset_index()

index volcano_name           country primary_volcano_type  \
0      6     Vesuvius             Italy        Stratovolcano   
1     13         Etna             Italy        Stratovolcano   
2    136   Nyiragongo          DR Congo        Stratovolcano   
3    281       Ulawun  Papua New Guinea        Stratovolcano   
4    379       Merapi         Indonesia        Stratovolcano   
5    864      Rainier     United States        Stratovolcano   
6    917    Mauna Loa     United States               Shield   
7   1057      Galeras          Colombia              Complex   

   activity_evidence last_known_eruption                          region  \
0  Eruption Observed             1944 CE  Mediterranean and Western Asia   
1  Eruption Observed             2022 CE  Mediterranean and Western Asia   
2  Eruption Observed             2022 CE              Africa and Red Sea   
3  Eruption Observed             2022 CE         Melanesia and Australia   
4  Eruption Observed             2022 CE                       Indonesia   
5     Eruption Dated             1450 CE          Canada and Western USA   
6  Eruption Observed             1984 CE        Hawaii and Pacific Ocean   
7  Eruption Observed             2014 CE                   South America   

          subregion  latitude  longitude  elevation_m  \
0             Italy    40.821     14.426         1281   
1             Italy    37.748     14.999         3357   
2  Africa (central)    -1.520     29.250         3470   
3       New Britain    -5.050    151.330         2334   
4              Java    -7.540    110.446         2910   
5  USA (Washington)    46.853   -121.760         4392   
6  Hawaiian Islands    19.475   -155.608         4170   
7          Colombia     1.220    -77.370         4276   

                   dominant_rock_type  \
0  Phono-tephrite /  Tephri-phonolite   
1    Trachybasalt / Tephrite Basanite   
2                             Foidite   
3               Basalt / Picro-Basalt   
4        Andesite / Basaltic Andesite   
5        Andesite / Basaltic Andesite   
6               Basalt / Picro-Basalt   
7        Andesite / Basaltic Andesite   

                               tectonic_setting  last_known_eruption_year  \
0  Subduction zone / Continental crust (>25 km)                    1944.0   
1  Subduction zone / Continental crust (>25 km)                    2022.0   
2        Rift zone / Continental crust (>25 km)                    2022.0   
3  Subduction zone / Continental crust (>25 km)                    2022.0   
4  Subduction zone / Continental crust (>25 km)                    2022.0   
5  Subduction zone / Continental crust (>25 km)                    1450.0   
6          Intraplate / Oceanic crust (< 15 km)                    1984.0   
7  Subduction zone / Continental crust (>25 km)                    2014.0   

  last_known_eruption_era  
0                      CE  
1                      CE  
2                      CE  
3                      CE  
4                      CE  
5                      CE  
6                      CE  
7                      CE

Vemos que la mitad de los volcanes ya coinciden. Para la otra mitad realizamos una búsqueda exhaustiva manual y creamos un diccionario con los nuevos nombres.

Sustituimos y validamos.

In [ ]:
rename_dict = {'Teide':'Tenerife',
'Avachinsky-Koryaksky': 'Avachinsky',
'Volcán de Colima': 'Colima',
'Santa María':'Santa Maria',
'Santorini (Thera)': 'Santorini',
 'Taal Volcano': 'Taal',
 'Sakurajima':'Aira',
 'Unzen': 'Unzendake' 
}
decade_volcanos.replace({'volcano':rename_dict}, inplace=True)
df_last_known_eruption[df_last_known_eruption['volcano_name'].isin(decade_volcanos.volcano)].reset_index()

index volcano_name           country primary_volcano_type  \
0       6     Vesuvius             Italy        Stratovolcano   
1      13         Etna             Italy        Stratovolcano   
2      19    Santorini            Greece               Shield   
3     136   Nyiragongo          DR Congo        Stratovolcano   
4     281       Ulawun  Papua New Guinea        Stratovolcano   
5     379       Merapi         Indonesia        Stratovolcano   
6     478         Taal       Philippines              Caldera   
7     506         Aira             Japan              Caldera   
8     510    Unzendake             Japan              Complex   
9     691   Avachinsky            Russia        Stratovolcano   
10    864      Rainier     United States        Stratovolcano   
11    917    Mauna Loa     United States               Shield   
12    954       Colima            Mexico        Stratovolcano   
13    976  Santa Maria         Guatemala        Stratovolcano   
14   1057      Galeras          Colombia              Complex   
15   1292     Tenerife             Spain        Stratovolcano   

    activity_evidence last_known_eruption                          region  \
0   Eruption Observed             1944 CE  Mediterranean and Western Asia   
1   Eruption Observed             2022 CE  Mediterranean and Western Asia   
2   Eruption Observed             1950 CE  Mediterranean and Western Asia   
3   Eruption Observed             2022 CE              Africa and Red Sea   
4   Eruption Observed             2022 CE         Melanesia and Australia   
5   Eruption Observed             2022 CE                       Indonesia   
6   Eruption Observed             2022 CE         Philippines and SE Asia   
7   Eruption Observed             2022 CE         Japan, Taiwan, Marianas   
8   Eruption Observed             1996 CE         Japan, Taiwan, Marianas   
9   Eruption Observed             2001 CE     Kamchatka and Mainland Asia   
10     Eruption Dated             1450 CE          Canada and Western USA   
11  Eruption Observed             1984 CE        Hawaii and Pacific Ocean   
12  Eruption Observed             2019 CE      México and Central America   
13  Eruption Observed             2022 CE      México and Central America   
14  Eruption Observed             2014 CE                   South America   
15  Eruption Observed             1909 CE                  Atlantic Ocean   

                    subregion  latitude  longitude  elevation_m  \
0                       Italy    40.821     14.426         1281   
1                       Italy    37.748     14.999         3357   
2                      Greece    36.404     25.396          367   
3            Africa (central)    -1.520     29.250         3470   
4                 New Britain    -5.050    151.330         2334   
5                        Java    -7.540    110.446         2910   
6                       Luzon    14.002    120.993          311   
7   Ryukyu Islands and Kyushu    31.593    130.657         1117   
8   Ryukyu Islands and Kyushu    32.761    130.299         1483   
9         Kamchatka Peninsula    53.256    158.836         2717   
10           USA (Washington)    46.853   -121.760         4392   
11           Hawaiian Islands    19.475   -155.608         4170   
12                     Mexico    19.514   -103.620         3850   
13                  Guatemala    14.757    -91.552         3745   
14                   Colombia     1.220    -77.370         4276   
15             Canary Islands    28.271    -16.641         3715   

                    dominant_rock_type  \
0   Phono-tephrite /  Tephri-phonolite   
1     Trachybasalt / Tephrite Basanite   
2                               Dacite   
3                              Foidite   
4                Basalt / Picro-Basalt   
5         Andesite / Basaltic Andesite   
6         Andesite / Basaltic Andesite   
7         Andesite / Basaltic Andesite   
8         Andesite / Basaltic Andesite   
9         Andesite / Basalti

### Uso de Merge

Ya con los dataset homogeneizados. Procedemos a fusionar las columnas relevantes de ambos.

In [ ]:
lat_long = df_last_known_eruption[[ 'volcano_name','latitude','longitude']]
decade_volcanos = pd.merge(decade_volcanos,lat_long,left_on='volcano',right_on='volcano_name')
decade_volcanos

volcano elevation                                     region  \
0    Avachinsky      2741                                  Kamchatka   
1        Colima      3820                             Colima/Jalisco   
2       Galeras      4276                                     Nariño   
3     Mauna Loa      4169                                     Hawaii   
4          Etna      3326                                     Sicily   
5        Merapi      2910  Central Java/Special Region of Yogyakarta   
6    Nyiragongo      3470                                 North Kivu   
7       Rainier      4392                                 Washington   
8      Vesuvius      1281                                   Campania   
9     Unzendake      1486                          Nagasaki/Kumamoto   
10         Aira      1117                                  Kagoshima   
11  Santa Maria      3772                             Quetzaltenango   
12    Santorini       367                               South Aegean   
13         Taal       311                                 Calabarzon   
14     Tenerife      3715                             Canary Islands   
15       Ulawun      2334          East New Britain/West New Britain   

                             country  pop_at_risk last_known_eruption  \
0                             Russia     179367.0                2001   
1                             Mexico     770000.0                2019   
2                           Colombia    1630000.0                2014   
3                      United States          0.0                1984   
4                              Italy    1100000.0                2013   
5                          Indonesia    5050000.0                2022   
6   Democratic Republic of the Congo    2000000.0                2021   
7                      United States     795000.0                1894   
8                              Italy    3085000.0                1944   
9                              Japan    1320000.0                1995   
10                             Japan    1470000.0                2022   
11                         Guatemala    1127000.0                2013   
12                            Greece          0.0                1950   
13                       Philippines   24393000.0                2021   
14                             Spain     900000.0                1909   
15                  Papua New Guinea          0.0                2019   

   volcano_name  latitude  longitude  
0    Avachinsky    53.256    158.836  
1        Colima    19.514   -103.620  
2       Galeras     1.220    -77.370  
3     Mauna Loa    19.475   -155.608  
4          Etna    37.748     14.999  
5        Merapi    -7.540    110.446  
6    Nyiragongo    -1.520     29.250  
7       Rainier    46.853   -121.760  
8      Vesuvius    40.821     14.426  
9     Unzendake    32.761    130.299  
10         Aira    31.593    130.657  
11  Santa Maria    14.757    -91.552  
12    Santorini    36.404     25.396  
13         Taal    14.002    120.993  
14     Tenerife    28.271    -16.641  
15       Ulawun    -5.050    151.330

In [ ]:
decade_volcanos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16 entries, 0 to 15
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   volcano              16 non-null     object 
 1   elevation            16 non-null     object 
 2   region               16 non-null     object 
 3   country              16 non-null     object 
 4   pop_at_risk          16 non-null     float64
 5   last_known_eruption  16 non-null     object 
 6   volcano_name         16 non-null     object 
 7   latitude             16 non-null     float64
 8   longitude            16 non-null     float64
dtypes: float64(3), object(6)
memory usage: 1.2+ KB


Ahora que tenemos toda la información necesaria, procedemos a realizar el útimo mapa.

Empezamos creando el mapa base y el objeto de agrupamiento.

In [ ]:
decade_volcanos_map = folium.Map(location=[20,-20],tiles='Stamen Terrain',zoom_start=2.5)
feature_group = folium.FeatureGroup()

Creamos las variables con la información que incluiremos en los marcadores.

In [ ]:
pop = decade_volcanos.pop_at_risk.apply(lambda x: str(x) if x !=0 else 'unknown')
radius = decade_volcanos.pop_at_risk.apply(lambda x: x/10 if x !=0 else 40000)
location = decade_volcanos.region + ', '+ decade_volcanos.country 
elevation = decade_volcanos.elevation.values
lat = decade_volcanos.latitude.values
lng = decade_volcanos.longitude.values
names = decade_volcanos.volcano.values

Creamos los marcadores y los agregamos al mapa.

In [ ]:
for lat,lng,name,location,radius,height,pop in zip(lat, lng,names,location,radius,elevation,pop):
  feature_group.add_child(
      folium.features.Marker(
          location=[lat,lng],
          tooltip= name,
          popup= location,
          icon= folium.features.CustomIcon(
              icon_image='https://volcano.si.edu/includes/images/volcano_icon_24.png',
              icon_size=(25,25))
      )
    )
  feature_group.add_child(
      folium.features.Circle(
          location=[lat,lng],
          radius=radius,
          color='orange',
          popup= 'Elevation: ' + height+'m',
          tooltip = 'Poppulation at risk: '+pop,
          fill= True
      )
  )


Y visualizamos nuestro último mapa.

In [ ]:
decade_volcanos_map.add_child(feature_group)
decade_volcanos_map

<font size=6 color='dodgerblue'>  8. Conclusiones.


---



Se identificó un problema que se pudo desarrollar adecuadamente para su exploración, se expusieron las razones para elegir el tema y se plantearon algunos cuestionamientos que pudieron ser resueltos en el desarrollo del proyecto mediate la colección de datos que se logró conseguir y explotar.
Se aplicaron las técnicas de Análisis Exploratorio de Datos y Limpieza de Datos lo que facilitó el desarrollo de la exploración.
Durante la exploración se tuvo que recurrir en diversas ocasiones a la generación de otros DataFrames de apoyo, así como efectuar diversas limpiezas de datos dado el propio avance y desarrollo del proyecto, si bien se siguió un orden en estos procesos, se tuvo que adecuar conforme la exploración avanzaba lo que permitió lograr el cometido propuesto.


##Hallazgos
De esta investigación podemos observar que hay dos regiones particularmente densas donde se agrupan la vasta mayoría de los volcanes del mundo las cuales corresponden al Cinturón de Fuego del Pacífico y al Rift de África Oriental.

Es de particular interés la región de Cinturón de Fuego del Pacífico que es a la que pertenece la geografía de México.

De hecho, como se vió México aparece en 9° posición en la densidad de volcanes por país, con lo que se ubica cuál es la situación de riesgo de nuestro país en el mundo.


##Respuestas al planteamiento inicial.

En relación a las preguntas planteadas para el desarrollo del proyecto estas se han respondido a lo largo del desarrollo del proyecto y, a manera de resumen, se tiene lo siguiente:

1.   ¿Cuáles volcanes representan mayor riesgo?
>    <font color='crimson'>**Respuesta**</font>: Son 16 los volcanes de mayor riesgo y fueron identificados con el **Mapa 3: Volcanes de Alto Riesgo (sección 7. Visualización de los datos: Mapas interactivos**).

>>1.   Mauna Loa (EU)
2.   Rainer (EU)
3.   Colima (México)
4.   Santa María (Guatemala)
5.   Galeras (Colombia)
6.   Tenerife (España)
7.   Vesuvius (Italia)
8.   Etna (Italia)
9.   Santorini (Gracia)
10.   Nyiragongo (República del Congo)
11.   Merapi (Indonesia)
12.   Taal (Filipinas)
13.   Ulawun (Papua Nueva Guinea)
14.   Aira (Japón)
15.   Unzebndake (Japón)
16.   Avachinsky (Rusia) 




2.   ¿Qué país tiene el mayor riesgo potencial a una catástrofe volcánica?
>    <font color='crimson'>**Respuesta**</font>: **Estados Unidos** por su densidad volcánica para un solo país.

3.   ¿De cuáles se tiene registro de -al menos una- erupción?
>    <font color='crimson'>**Respuesta**</font>: **Son 1,329**.
> La respuesta se ubica en **Eliminación de NaNs (6. Transformación, filtrado y ordenamiento de datos. - División del Tiempo)**

4.   ¿Qué tipos de volcanes existen?
>    <font color='crimson'>**Respuesta**</font>: Hay 19 tipos de volcanes, siendo el tipo **Stratovolcano** el de mayor cantidad con 664 que equivale a un 50%.

In [ ]:
volcano_types_df['pct'] = (volcano_types_df['count']/volcano_types_df['count'].sum())*100 
volcano_types_df

primary_volcano_type  count        pct  avg_elevation
0         Stratovolcano    664  49.849850    2203.597892
1                Shield    124   9.309309    1509.137097
2             Submarine    116   8.708709    -909.637931
3       Pyroclasticcone     96   7.207207    1660.760417
4               Caldera     80   6.006006    1475.450000
5         Volcanicfield     77   5.780781    1572.402597
6               Complex     53   3.978979    2290.603774
7              Lavadome     37   2.777778    1646.702703
8           Fissurevent     19   1.426426     973.473684
9                  Maar     15   1.126126    1238.733333
10             Compound     13   0.975976    2600.000000
11             Tuffcone     12   0.900901    1221.500000
12    Pyroclasticshield      8   0.600601    2413.000000
13           Craterrows      6   0.450450     526.500000
14           Subglacial      5   0.375375    1518.000000
15             Lavacone      4   0.300300    2129.750000
16      Explosioncrater      1   0.075075     -48.000000
17                 Cone      1   0.075075    1326.000000
18             Tuffring      1   0.075075     830.000000

5.   ¿Dónde se ubican? (uso de mapas interactivos)
>    <font color='crimson'>**Respuesta**</font>:  Los volcanes fueron identificados con el **Mapa 1: Volcanes del Mundo (sección 7. Visualización de los datos: Mapas interactivos**).


6.   ¿Qué país/continente/región tiene la mayor cantidad de volcanes?
>    <font color='crimson'>**Respuesta**</font>: La respuesta se obtiene al navegar en el **Mapa 1: Volcanes del Mundo (sección 7. Visualización de los datos: Mapas interactivos**) 
>*   País: **Estados Unidos con 162**
>*   Continente: **Americano con 515** 
>*   Región: **South America con 176**




7.   Altitud media.
>    <font color='crimson'>**Respuesta**</font>: 
>- (a) 1,932.6 metros sobre el nivel del mar; 
>- (b) 1,690.6 metros considerando altitud absoluta


8.   Relaciones tectónica de placas - tipo de volcán.
>    <font color='crimson'>**Respuesta**</font>:

In [ ]:
tectonic_setting[['tectonic_setting','primary_volcano_type']].groupby(['tectonic_setting','primary_volcano_type']).size().head(40)

tectonic_setting                            primary_volcano_type
Intraplate / Continental crust (>25 km)     Caldera                  2
                                            Fissurevent              1
                                            Lavacone                 1
                                            Lavadome                 1
                                            Maar                     1
                                            Pyroclasticcone         39
                                            Pyroclasticshield        1
                                            Shield                  22
                                            Stratovolcano           31
                                            Subglacial               1
                                            Submarine                2
                                            Tuffcone                 1
                                            Volcanicfield           31
Intraplate /